bazel/repo.bzl

```diff
@@ -0,0 +1,46 @@
+def _envoy_repo_impl(repository_ctx):
+    """This provides information about the Envoy repository
+
+    You can access the current version and path to the repository in .bzl/BUILD
+    files as follows:
+
+    ```starlark
+
+    load("@envoy_repo//:version.bzl", "VERSION")
+    load("@envoy_repo//:path.bzl", "PATH")
```

Comment:
What do you plan to use this for? I'm afraid this is an anti-pattern because wherever you pull this in rules/targets it invalidates cache across different builds.

Reply:
for tooling such as code_format where it needs to run across the whole repo, or where it needs to make bazel queries which imply access to the repo (ie not genqueriable)

i get that using the path in bazel is a bit of an anti-pattern, not least because its only relevant to the local machine, but we have quite a few rules that rely on having access to the repo, and currently that has to be passed around (as an absolute url), so i think from a pattern pov this would be preferable to what we do now

im not sure if iiuc the point about invalidating caches, but i dont think running dep check or code formatting tools is going to have any negative lateral effects, and afaict using this with them everything is cached correctly

Reply:
as an example - here is some WIP code to run a bazel query from other tools:

https://github.com/envoyproxy/envoy/pull/17992/files#diff-7a7d58af3edfdbe7825cd20fb6a29494ae293d40fe25e2b9774a4aac8ee27dcfR22-R30

the tools in this PR either dont currently use bazel - ie they are currently run as scripts

if i shift them to bazel then, like with several others, we need to add a path to be able to run the bazel queries

the tools that need a path are a pita to run as you always have to type `${PWD}` in addtion to any args, and it increases their code/complexity having to handle the path arg. This would eliminate that for these cases

Reply:
perhaps i could add something to the comment to clarify any situations where there are some adverse effect from using the `PATH` other than for rbe

Reply:
My point is this should be exported in a `.bzl` file as Starlark, for the tooling we only need the `py_library`?

Reply:
theres no way to get the path without this i think

Reply:
(im sure i could be wrong on this, but i dont know of any other way to get the local repo path)

Reply:
> we only need the py_library

if iiuc this point, no, we also need the provider for the py library to give it the path

see here https://github.com/envoyproxy/envoy/pull/17992/files#diff-7a7d58af3edfdbe7825cd20fb6a29494ae293d40fe25e2b9774a4aac8ee27dcfR15

Reply:
trying to understand further what you might mean here.

im wondering if you mean we dont need the ".bzl" part for the PATH, but we could keep the python part and only expose it to python targets

as im working mostly with python that would wfm

ill update with this change and then hopefully (if that was your meaning) we can get this landed today

Reply:
Have you considered the trick we do in https://github.com/envoyproxy/envoy/blob/68f674eb1364688d1aa1b2b41580bfd56cb844f7/tools/config_validation/validate_fragment.py#L71?

Reply:
that gives the path within the sandbox, not the repo, iiuc

Reply:
ah, i see, you can get the file path from the symlink that way

tbh, i think this is much more explicit - retrieving the repo metadata will be possible with the `@envoy_repo` target - which can provide version as well as path

see https://github.com/envoyproxy/envoy/pull/18021/files#diff-615ebb7b91831a9c7da2a3835e8b7c1ef89c9d2e72777cc229e128f6c8229214R51 and https://github.com/envoyproxy/envoy/pull/18021/files#diff-7a7d58af3edfdbe7825cd20fb6a29494ae293d40fe25e2b9774a4aac8ee27dcfR22 for an example

possibly the `envoy_repo` namespace can also provide data mappings as in #17998

Reply:
I like the idea of `@envoy_repo` for things like version, I just wonder if we can pair it back to that and we don't need to do path this way.